### PR TITLE
Update stats object to 1.1 spec

### PR DESCRIPTION
**Pull request by gernest**
This implements

- TS1
- TS2
- TS3
- TS4
- TS5
- TS6
- TS7
- TS8
- TS9

### CODE REVIEW

**Comment by gernest** [comment]
@paddybyers @ORBAT I think this is ready for review.

File: ably/proto/stat.go [hunk 1]
```diff
@@ -69,3 +21,125 @@ var (
...
+	Outbound MessageTraffic `json:"outbound" codec:"outbound"`
+}
+
+type XchgMessages struct {
+	All          MessageTypes `json:"all" codec:"all"`
+	ProducerPaid MessageTypes `json:"producerPaid" codec:"producerPaid"`
+	ConsumerPaid MessageTypes `json:"consumerPaid" codec:"consumerPaid"`
+}
+
+type BaseStats struct {
```
Comment (paddybyers) [inline ably/proto/stat.go]: This is superflous here, unless you can make `Stats` extend `BaseStats` and remove the common fields.

**Comment by SimonWoolf** [comment]
> what's your view on incuding the exchange-related stats types in the libraries? I'm inclined to include them while we're making updates, unless we're not confident that the structs are stable

Happy for them to be included, unless we want to start telling people to use untyped/flattened stats if they want exchange stats (and add support for that to client libs), per https://github.com/ably/wiki/issues/177

File: ably/proto/stat.go [hunk 2]
```diff
@@ -69,3 +21,125 @@ var (
...
+type MessageDirections struct {
+	All      MessageTypes   `json:"all" codec:"all"`
+	Inbound  MessageTraffic `json:"inbound" codec:"inbound"`
+	Outbound MessageTraffic `json:"outbound" codec:"outbound"`
+}
+
+type XchgMessages struct {
+	All          MessageTypes `json:"all" codec:"all"`
+	ProducerPaid MessageTypes `json:"producerPaid" codec:"producerPaid"`
+	ConsumerPaid MessageTypes `json:"consumerPaid" codec:"consumerPaid"`
```
Comment (SimonWoolf) [inline ably/proto/stat.go]: producerpaid and consumerpaid are messagedirections, not messagetypes

File: ably/proto/stat.go (hunk 1 above)
Comment (gernest) [reply ably/proto/stat.go]: my bad, will remove that .

**Comment by gernest** [comment]
@SimonWoolf @paddybyers  I addressed  the comments. Please let me know if there is a need for more changes, thanks.

**Comment by mattheworiordan** [comment]
>  what's your view on incuding the exchange-related stats types in the libraries? I'm inclined to include them while we're making updates, unless we're not confident that the structs are stable

I am dead against this.  This creates a lot more work for developers unnecessarily, given we have no evidence anyone is using the stats from the libs, and more importantly, the libs have no way of getting account level stats anyway.

I would prefer we instead added a new beta API perhaps that allowed access to untyped/flattened stats, and moved us closer towards an API we intend on supporting moving forwards.

**Pull request merged**